src/Microsoft.AspNet.Mvc.Core/Internal/ValidatorCache.cs

```diff
@@ -0,0 +1,144 @@
+            {
+                return GetValidatorsFromEntry(entry, metadata, validatorProvider);
+            }
+
+            var items = new List<ValidatorItem>(metadata.ValidatorMetadata.Count);
+            for (var i = 0; i < metadata.ValidatorMetadata.Count; i++)
+            {
+                items.Add(new ValidatorItem(metadata.ValidatorMetadata[i]));
+            }
+
+            ExecuteProvider(validatorProvider, metadata, items);
+
+            var validators = ExtractValidators(items);
+
+            var allValidatorsCached = true;
+            for (var i = 0; i < items.Count; i++)
+            {
+                var item = items[i];
+                if (!item.IsReusable)
+                {
+                    item.Validator = null;
+                    allValidatorsCached = false;
+                }
+            }
+
+            if (allValidatorsCached)
+            {
+                entry = new CacheEntry(validators);
+            }
+            else
+            {
+                entry = new CacheEntry(items);
+            }
+
+            _cacheEntries.TryAdd(metadata, entry);
+
+            return validators;
+        }
+
+        private IReadOnlyList<IModelValidator> GetValidatorsFromEntry(CacheEntry entry, ModelMetadata metadata, IModelValidatorProvider validationProvider)
+        {
+            Debug.Assert(entry.Validators != null || entry.Items != null);
+
+            if (entry.Validators != null)
+            {
+                return entry.Validators;
+            }
+
+            var items = new List<ValidatorItem>(entry.Items.Count);
+            for (var i = 0; i < entry.Items.Count; i++)
+            {
+                var item = entry.Items[i];
+                if (item.IsReusable)
+                {
+                    items.Add(item);
+                }
+                else
+                {
+                    items.Add(new ValidatorItem(item.ValidatorMetadata));
+                }
+            }
+
+            ExecuteProvider(validationProvider, metadata, items);
+
+            return ExtractValidators(items);
+        }
+
+        private void ExecuteProvider(IModelValidatorProvider validatorProvider, ModelMetadata metadata, List<ValidatorItem> items)
+        {
+            var context = new ModelValidatorProviderContext(metadata, items);
+
+            validatorProvider.GetValidators(context);
+        }
+
+        private IReadOnlyList<IModelValidator> ExtractValidators(List<ValidatorItem> items)
+        {
+            var count = 0;
+            for (var i = 0; i < items.Count; i++)
+            {
+                if (items[i].Validator != null)
+                {
+                    count++;
+                }
+            }
+
+            if (count == 0)
+            {
+                return EmptyArray;
+            }
+
+            var validators = new IModelValidator[count];
+            for (int i = 0, j = 0; i < items.Count; i++)
```

Comment:
I kind of agree with what you are saying but having a variable declaration like `var j = 0` outside a for loop looks odd to me.